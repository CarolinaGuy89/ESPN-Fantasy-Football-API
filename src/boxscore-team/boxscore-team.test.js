import _ from 'lodash';

import BaseObject from '../base-object/base-object.js';
import Team from '../team/team.js';

import BoxscoreTeam from './boxscore-team.js';

import { localObject, serverResponse } from './boxscore-team.stubs.js';

describe('BoxscoreTeam', () => {
  let boxscoreTeam;

  beforeEach(() => {
    boxscoreTeam = new BoxscoreTeam({ leagueId: 234234, seasonId: 2017 });
  });

  afterEach(() => {
    boxscoreTeam = null;
  });

  test('extends BaseObject', () => {
    expect(boxscoreTeam).toBeInstanceOf(BaseObject);
  });

  describe('attribute population from server response', () => {
    beforeEach(() => {
      boxscoreTeam = BoxscoreTeam.buildFromServer(serverResponse);
    });

    test('parses data correctly', () => {
      expect(boxscoreTeam).toMatchSnapshot();
    });
  });

  describe('attribute population from local object', () => {
    beforeEach(() => {
      boxscoreTeam = new BoxscoreTeam(localObject);
    });

    test('parses data correctly', () => {
      expect(boxscoreTeam).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    describe('when options are not passed', () => {
      const testPropIsUndefined = (prop) => {
        test(`${prop} is undefined`, () => {
          const newInstance = new BoxscoreTeam();
          expect(_.get(newInstance, prop)).toBeUndefined();
        });
      };

      testPropIsUndefined('leagueId');
      testPropIsUndefined('seasonId');
    });

    describe('when options are passed', () => {
      const testPropIsSetFromOptions = (prop) => {
        test(`${prop} is set from options`, () => {
          const value = 25;
          const newInstance = new BoxscoreTeam({ [prop]: value });
          expect(_.get(newInstance, prop)).toBe(value);
        });
      };

      testPropIsSetFromOptions('leagueId');
      testPropIsSetFromOptions('seasonId');
    });
  });

  describe('responseMap', () => {
    describe('team', () => {
      describe('manualParse', () => {
        describe('when the passed data is undefined', () => {
          test('returns undefined', () => {
            const returnedTeam = BoxscoreTeam.responseMap.team.manualParse();
            expect(returnedTeam).toBeUndefined();
          });
        });

        describe('when the passed data is empty', () => {
          test('returns undefined', () => {
            const returnedTeam = BoxscoreTeam.responseMap.team.manualParse({});
            expect(returnedTeam).toBeUndefined();
          });
        });

        describe('when the passed data is populated', () => {
          let ids, responseData, teamId;

          beforeEach(() => {
            Team.clearCache();

            teamId = 10;
            responseData = {
              teamId,
              team: { teamId }
            };
            ids = {
              leagueId: boxscoreTeam.leagueId,
              seasonId: boxscoreTeam.seasonId,
              teamId
            };
          });

          afterEach(() => {
            ids = null;
            responseData = null;
            teamId = null;

            Team.clearCache();
          });

          describe('when there is a cached team', () => {
            test('returns the cached team', () => {
              const cachedTeam = Team.buildFromServer({}, ids);

              const returnedTeam = BoxscoreTeam.responseMap.team.manualParse(
                responseData, undefined, boxscoreTeam
              );
              expect(returnedTeam).toBe(cachedTeam);
            });
          });

          describe('when there is not a cached team', () => {
            test('creates a new team', () => {
              const cachingId = Team.getCacheId(ids);

              const returnedTeam = BoxscoreTeam.responseMap.team.manualParse(
                responseData, undefined, boxscoreTeam
              );

              expect(returnedTeam).toBe(Team.get(cachingId));
            });
          });
        });
      });
    });
  });
});
