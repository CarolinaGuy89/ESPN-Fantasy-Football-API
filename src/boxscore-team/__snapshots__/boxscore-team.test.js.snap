// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxscoreTeam attribute population from local object parses data correctly 1`] = `
BoxscoreTeam {
  "leagueId": undefined,
  "matchupScore": 204,
  "players": Array [
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "QB",
          "Bench",
        ],
        "firstName": "Aaron",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "12",
        "lastName": "Rodgers",
        "percentOwned": 99.21669,
        "percentOwnedChange": 0.23744,
        "percentStarted": 74.42904,
        "playerId": 8439,
      },
      "position": "QB",
      "projectedPoints": 20.46,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.47,
        "passing2PtConversions": 0.19,
        "passingInterceptions": -0.76,
        "passingTouchdowns": 7.87,
        "passingYards": 11.21,
        "rushing2PtConversions": 0.01,
        "rushingTouchdowns": 0.67,
        "rushingYards": 1.73,
      },
      "stats": BoxscorePlayerPointStats {
        "passingYards": 1.04,
      },
      "totalPoints": 1.04,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "RB",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Mark",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "22",
        "lastName": "Ingram II",
        "percentOwned": 94.37537,
        "percentOwnedChange": -1.14045,
        "percentStarted": 30.44836,
        "playerId": 13981,
      },
      "position": "RB",
      "projectedPoints": 5.67,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.19,
        "receiving2PtConversions": 0.01,
        "receivingTouchdowns": 0.27,
        "receivingYards": 0.92,
        "rushing2PtConversions": 0.02,
        "rushingTouchdowns": 1.49,
        "rushingYards": 3.13,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 1.5,
        "rushingYards": 2.8,
      },
      "totalPoints": 4.3,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "RB",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Tevin",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "26",
        "lastName": "Coleman",
        "percentOwned": 91.59903,
        "percentOwnedChange": -0.95381,
        "percentStarted": 42.36213,
        "playerId": 17689,
      },
      "position": "RB",
      "projectedPoints": 10.41,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.13,
        "receiving2PtConversions": 0.01,
        "receivingTouchdowns": 0.96,
        "receivingYards": 2.41,
        "rushing2PtConversions": 0.04,
        "rushingTouchdowns": 2.09,
        "rushingYards": 5.03,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 0.9,
        "rushingTouchdowns": 6,
        "rushingYards": 4.5,
      },
      "totalPoints": 11.4,
    },
  ],
  "seasonId": undefined,
  "team": Team {
    "abbreviation": "NUTS",
    "division": Object {
      "divisionId": 0,
      "divisionName": "Division 1",
      "size": 4,
    },
    "divisionStanding": 4,
    "firstName": "No Bell",
    "lastName": "Peace Prize",
    "leagueId": undefined,
    "leagueStanding": 6,
    "logoURL": "https://i.imgur.com/OQhSYj1.gif",
    "seasonId": undefined,
    "streakType": "ERROR: streakType not recognized",
    "teamId": 9,
    "waiverRank": 1,
  },
  "weekBenchProjectedScore": 47.56,
  "weekBenchScore": 72.06,
  "weekProjectedScore": 90.47,
  "weekScore": 70.64,
}
`;

exports[`BoxscoreTeam attribute population from server response parses data correctly 1`] = `
BoxscoreTeam {
  "leagueId": undefined,
  "matchupScore": 234,
  "players": Array [
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "QB",
          "Bench",
        ],
        "firstName": "Aaron",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "12",
        "lastName": "Rodgers",
        "percentOwned": 99.21669,
        "percentOwnedChange": 0.23744,
        "percentStarted": 74.42904,
        "playerId": 8439,
      },
      "position": "QB",
      "projectedPoints": 20.46,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.47,
        "passing2PtConversions": 0.19,
        "passingInterceptions": -0.76,
        "passingTouchdowns": 7.87,
        "passingYards": 11.21,
        "rushing2PtConversions": 0.01,
        "rushingTouchdowns": 0.67,
        "rushingYards": 1.73,
      },
      "stats": BoxscorePlayerPointStats {
        "passingYards": 1.04,
      },
      "totalPoints": 1.04,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "RB",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Mark",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "22",
        "lastName": "Ingram II",
        "percentOwned": 94.37537,
        "percentOwnedChange": -1.14045,
        "percentStarted": 30.44836,
        "playerId": 13981,
      },
      "position": "RB",
      "projectedPoints": 5.67,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.19,
        "receiving2PtConversions": 0.01,
        "receivingTouchdowns": 0.27,
        "receivingYards": 0.92,
        "rushing2PtConversions": 0.02,
        "rushingTouchdowns": 1.49,
        "rushingYards": 3.13,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 1.5,
        "rushingYards": 2.8,
      },
      "totalPoints": 4.3,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "RB",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Tevin",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "26",
        "lastName": "Coleman",
        "percentOwned": 91.59903,
        "percentOwnedChange": -0.95381,
        "percentStarted": 42.36213,
        "playerId": 17689,
      },
      "position": "RB",
      "projectedPoints": 10.41,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.13,
        "receiving2PtConversions": 0.01,
        "receivingTouchdowns": 0.96,
        "receivingYards": 2.41,
        "rushing2PtConversions": 0.04,
        "rushingTouchdowns": 2.09,
        "rushingYards": 5.03,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 0.9,
        "rushingTouchdowns": 6,
        "rushingYards": 4.5,
      },
      "totalPoints": 11.4,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "WR",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Julio",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "11",
        "lastName": "Jones",
        "percentOwned": 99.93748,
        "percentOwnedChange": -0.00715,
        "percentStarted": 97.29097,
        "playerId": 13982,
      },
      "position": "WR",
      "projectedPoints": 13.65,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.1,
        "receiving2PtConversions": 0.06,
        "receivingTouchdowns": 3.91,
        "receivingYards": 9.61,
        "rushing2PtConversions": 0,
        "rushingTouchdowns": 0.03,
        "rushingYards": 0.13,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingTouchdowns": 6,
        "receivingYards": 13.8,
      },
      "totalPoints": 19.8,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "WR",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "JuJu",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "19",
        "lastName": "Smith-Schuster",
        "percentOwned": 99.27393,
        "percentOwnedChange": 0.09938,
        "percentStarted": 96.47065,
        "playerId": 18306,
      },
      "position": "WR",
      "projectedPoints": 12.5,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.06,
        "receiving2PtConversions": 0.07,
        "receivingTouchdowns": 4.03,
        "receivingYards": 8.46,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingTouchdowns": 6,
        "receivingYards": 3.7,
      },
      "totalPoints": 9.7,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "TE",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "David",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "85",
        "lastName": "Njoku",
        "percentOwned": 72.99284,
        "percentOwnedChange": -1.10941,
        "percentStarted": 38.99002,
        "playerId": 18303,
      },
      "position": "TE",
      "projectedPoints": 5.26,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.04,
        "receiving2PtConversions": 0.04,
        "receivingTouchdowns": 1.42,
        "receivingYards": 3.84,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 4.2,
      },
      "totalPoints": 4.2,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "RB",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Tarik",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "29",
        "lastName": "Cohen",
        "percentOwned": 93.2453,
        "percentOwnedChange": -0.48103,
        "percentStarted": 56.8237,
        "playerId": 18171,
      },
      "position": "RB/WR/TE",
      "projectedPoints": 8.92,
      "projectedStats": BoxscorePlayerPointStats {
        "kickoffReturnTouchdown": 0,
        "lostFumbles": -0.14,
        "puntReturnTouchdown": 0.16,
        "receiving2PtConversions": 0.02,
        "receivingTouchdowns": 1.12,
        "receivingYards": 4.53,
        "rushing2PtConversions": 0.01,
        "rushingTouchdowns": 0.69,
        "rushingYards": 2.53,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 0.8,
        "rushingTouchdowns": 6,
        "rushingYards": 2.4,
      },
      "totalPoints": 9.2,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "D/ST",
          "Bench",
        ],
        "firstName": "Bears",
        "isActive": true,
        "isIREligible": false,
        "lastName": "D/ST",
        "percentOwned": 97.27124,
        "percentOwnedChange": -0.48172,
        "percentStarted": 75.43181,
        "playerId": 60003,
      },
      "position": "D/ST",
      "projectedPoints": 6.46,
      "projectedStats": BoxscorePlayerPointStats {
        "defensiveBlockedKickForTouchdowns": 0.05,
        "defensiveBlockedKicks": 0.21,
        "defensiveFumbles": 1.12,
        "defensiveInterceptions": 1.72,
        "defensiveSacks": 2.51,
        "defensiveSafeties": 0.04,
        "fumbleReturnTouchdown": 0.12,
        "interceptionReturnTouchdown": 0.47,
        "kickoffReturnTouchdown": 0.06,
        "puntReturnTouchdown": 0.16,
      },
      "stats": BoxscorePlayerPointStats {
        "defensive100To199YardsAllowed": 3,
        "defensive7To13PointsAllowed": 3,
        "defensiveSacks": 3,
      },
      "totalPoints": 9,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "K",
          "Bench",
        ],
        "firstName": "Wil",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "3",
        "lastName": "Lutz",
        "percentOwned": 95.81607,
        "percentOwnedChange": -0.73695,
        "percentStarted": 87.27326,
        "playerId": 18104,
      },
      "position": "K",
      "projectedPoints": 7.15,
      "projectedStats": BoxscorePlayerPointStats {
        "madeExtraPoints": 2.46,
        "madeFieldGoalsFrom40To49": 1.63,
        "madeFieldGoalsFrom50Plus": 0.82,
        "madeFieldGoalsFromUnder40": 2.58,
        "missedExtraPoints": -0.11,
        "missedFieldGoals": -0.22,
      },
      "stats": BoxscorePlayerPointStats {
        "madeExtraPoints": 2,
      },
      "totalPoints": 2,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "RB",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Jordan",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "24",
        "lastName": "Howard",
        "percentOwned": 88.57899,
        "percentOwnedChange": -0.73251,
        "percentStarted": 33.90609,
        "playerId": 17919,
      },
      "position": "Bench",
      "projectedPoints": 7.95,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.11,
        "receiving2PtConversions": 0,
        "receivingTouchdowns": 0.2,
        "receivingYards": 0.74,
        "rushing2PtConversions": 0.03,
        "rushingTouchdowns": 1.74,
        "rushingYards": 5.35,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 0.5,
        "rushingTouchdowns": 12,
        "rushingYards": 10.9,
      },
      "totalPoints": 23.4,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "WR",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Alshon",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "17",
        "lastName": "Jeffery",
        "percentOwned": 92.4714,
        "percentOwnedChange": 0.05134,
        "percentStarted": 54.13407,
        "playerId": 14912,
      },
      "position": "Bench",
      "projectedPoints": 7.71,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.05,
        "receiving2PtConversions": 0.04,
        "receivingTouchdowns": 2.05,
        "receivingYards": 5.68,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingTouchdowns": 6,
        "receivingYards": 5.9,
      },
      "totalPoints": 11.9,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "TE",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Kyle",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "82",
        "lastName": "Rudolph",
        "percentOwned": 82.67324,
        "percentOwnedChange": 2.35959,
        "percentStarted": 48.47325,
        "playerId": 14054,
      },
      "position": "Bench",
      "projectedPoints": 6.3,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.03,
        "receiving2PtConversions": 0.04,
        "receivingTouchdowns": 2.15,
        "receivingYards": 4.14,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingYards": 1.9,
      },
      "totalPoints": 1.9,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "WR",
          "RB/WR/TE",
          "Bench",
        ],
        "firstName": "Calvin",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "18",
        "lastName": "Ridley",
        "percentOwned": 82.24974,
        "percentOwnedChange": 1.19766,
        "percentStarted": 31.08707,
        "playerId": 18584,
      },
      "position": "Bench",
      "projectedPoints": 7.03,
      "projectedStats": BoxscorePlayerPointStats {
        "kickoffReturnTouchdown": 0.01,
        "lostFumbles": -0.06,
        "receiving2PtConversions": 0.03,
        "receivingTouchdowns": 1.94,
        "receivingYards": 4.79,
        "rushing2PtConversions": 0,
        "rushingTouchdowns": 0.07,
        "rushingYards": 0.25,
      },
      "stats": BoxscorePlayerPointStats {
        "receivingTouchdowns": 6,
        "receivingYards": 3.2,
      },
      "totalPoints": 9.2,
    },
    BoxscorePlayer {
      "isLocked": true,
      "player": Player {
        "eligiblePositions": Array [
          "QB",
          "Bench",
        ],
        "firstName": "Lamar",
        "isActive": true,
        "isIREligible": false,
        "jerseyNumber": "8",
        "lastName": "Jackson",
        "percentOwned": 57.39529,
        "percentOwnedChange": 5.06828,
        "percentStarted": 24.54791,
        "playerId": 18590,
      },
      "position": "Bench",
      "projectedPoints": 18.56,
      "projectedStats": BoxscorePlayerPointStats {
        "lostFumbles": -0.77,
        "passing2PtConversions": 0.15,
        "passingInterceptions": -1.25,
        "passingTouchdowns": 4.39,
        "passingYards": 6.62,
        "rushing2PtConversions": 0.02,
        "rushingTouchdowns": 2.65,
        "rushingYards": 6.72,
      },
      "stats": BoxscorePlayerPointStats {
        "lostFumbles": -2,
        "passingYards": 7.16,
        "rushingTouchdowns": 12,
        "rushingYards": 8.5,
      },
      "totalPoints": 25.66,
    },
    BoxscorePlayer {
      "position": "Bench",
      "projectedStats": BoxscorePlayerPointStats {},
      "stats": BoxscorePlayerPointStats {},
    },
    BoxscorePlayer {
      "position": "Bench",
      "projectedStats": BoxscorePlayerPointStats {},
      "stats": BoxscorePlayerPointStats {},
    },
  ],
  "seasonId": undefined,
  "team": Team {
    "abbreviation": "NUTS",
    "division": Object {
      "divisionId": 0,
      "divisionName": "Division 1",
      "size": 4,
    },
    "divisionStanding": 4,
    "firstName": "No Bell",
    "lastName": "Peace Prize",
    "leagueId": undefined,
    "leagueStanding": 6,
    "logoURL": "https://i.imgur.com/OQhSYj1.gif",
    "seasonId": undefined,
    "streakType": "ERROR: streakType not recognized",
    "teamId": 9,
    "waiverRank": 1,
  },
  "weekBenchProjectedScore": 47.56,
  "weekBenchScore": 72.06,
  "weekProjectedScore": 90.47,
  "weekScore": 70.64,
}
`;
